src/nvim/lua/treesitter.c

```diff
@@ -1361,173 +1365,133 @@ static int node_equal(lua_State *L)
 
-  if (lua_gettop(L) >= 4) {
-    uint32_t start = (uint32_t)luaL_checkinteger(L, 4);
-    uint32_t end = lua_gettop(L) >= 5 ? (uint32_t)luaL_checkinteger(L, 5) : MAXLNUM;
-    ts_query_cursor_set_point_range(cursor, (TSPoint){ start, 0 }, (TSPoint){ end, 0 });
+  TSQueryMatch match;
+  uint32_t capture_index;
+  if (!ts_query_cursor_next_capture(cursor, &match, &capture_index)) {
+    return 0;
   }
 
-  if (lua_gettop(L) >= 6 && !lua_isnil(L, 6)) {
-    if (!lua_istable(L, 6)) {
-      return luaL_error(L, "table expected");
-    }
-    lua_pushnil(L);
-    // stack: [dict, ..., nil]
-    while (lua_next(L, 6)) {
-      // stack: [dict, ..., key, value]
-      if (lua_type(L, -2) == LUA_TSTRING) {
-        char *k = (char *)lua_tostring(L, -2);
-        if (strequal("max_start_depth", k)) {
-          uint32_t max_start_depth = (uint32_t)lua_tointeger(L, -1);
-          ts_query_cursor_set_max_start_depth(cursor, max_start_depth);
-        }
-      }
-      lua_pop(L, 1);  // pop the value; lua_next will pop the key.
-      // stack: [dict, ..., key]
-    }
-  }
+  TSQueryCapture capture = match.captures[capture_index];
 
-  TSLua_cursor *ud = lua_newuserdata(L, sizeof(*ud));  // [udata]
-  ud->cursor = cursor;
-  ud->predicated_match = -1;
-  ud->max_match_id = -1;
+  // Handle capture quantifiers here
+  lua_pushinteger(L, capture.index + 1);  // [index]
+  push_node(L, capture.node, 1);  // [index, node]
+  push_querymatch(L, &match, 1);
 
-  lua_getfield(L, LUA_REGISTRYINDEX, TS_META_QUERYCURSOR);
-  lua_setmetatable(L, -2);  // [udata]
-  lua_pushvalue(L, 1);  // [udata, node]
+  return 3;
+}
 
-  // include query separately, as to keep a ref to it for gc
-  lua_pushvalue(L, 2);  // [udata, node, query]
+static int querycursor_next_match(lua_State *L)
+{
+  TSQueryCursor *cursor = querycursor_check(L, 1);
 
-  if (captures) {
-    // placeholder for match state
-    lua_createtable(L, (int)ts_query_capture_count(query), 2);  // [u, n, q, match]
-    lua_pushcclosure(L, query_next_capture, 4);  // [closure]
-  } else {
-    lua_pushcclosure(L, query_next_match, 3);  // [closure]
+  TSQueryMatch match;
+  if (!ts_query_cursor_next_match(cursor, &match)) {
+    return 0;
   }
 
+  push_querymatch(L, &match, 1);
+
   return 1;
 }
 
+static TSQueryCursor *querycursor_check(lua_State *L, int index)
+{
+  TSQueryCursor **ud = luaL_checkudata(L, index, TS_META_QUERYCURSOR);
+  return *ud;
+}
+
 static int querycursor_gc(lua_State *L)
 {
-  TSLua_cursor *ud = luaL_checkudata(L, 1, TS_META_QUERYCURSOR);
-  kv_push(cursors, ud->cursor);
-  ud->cursor = NULL;
+  TSQueryCursor *cursor = querycursor_check(L, 1);
+  ts_query_cursor_delete(cursor);
   return 0;
 }
 
+static int querymatch_info(lua_State *L)
+{
+  TSQueryMatch *ud = luaL_checkudata(L, 1, TS_META_QUERYMATCH);
+  lua_pushinteger(L, ud->id);
+  lua_pushinteger(L, ud->pattern_index + 1);
+  return 2;
+}
+
+static int querymatch_captures(lua_State *L)
```

Comment:
Keeping this out of `querymatch_info` means paying the cost of building the captures table is opt-in. A future improvement will be to cache this table to speed up multiple calls to `match:captures()`.